### DISCUSSION TITLE
Using remark-retext with MDX (in Next.js).

### DISCUSSION THREAD

**Discussion by sebastienbarre** [remarkjs/remark#965] [Q&A] [answered]
I was wondering if you could provide guidance with respect to using [retextjs/retext](https://github.com/retextjs/retext) with remark in the context of a MDX file and Next.js.

Next.js supports MDX either out-of-the-box through [@next/mdx](https://nextjs.org/docs/advanced-features/using-mdx#setup-nextmdx-in-nextjs) or via [hashicorp/next-mdx-remote](https://github.com/hashicorp/next-mdx-remote#apis). Both paths allow remark and rehype plugins to be specified as options:

```
 options: {
    remarkPlugins: [],
    rehypePlugins: [],
  },
```

I found out about [remarkjs/remark-retext: plugin to transform from remark (Markdown) to retext (natural language)](https://github.com/remarkjs/remark-retext), which I assume could be slotted into the `remarkPlugins` array.

However, retext works through plugins and I'm a bit lost as to *how* I can pass a list of retext plugins to `options`, or to `remark-text`. 

Thank you

**Comment by ChristianMurphy** [comment]
@sebastienbarre could you expand a bit on what your goal is?

retext can be used in a number of ways, naming just a few:
* extracting plain text from markdown without the formatting or components
* spelling or grammar checking markdown, without viewing the rendered content
* spelling or grammar checking markdown, while also viewing the rendered content
* and many more

the appropriate way to use it depends on the goal.

and will influence which way `mdx` should be called, what mode `remark-retext` should be run in (https://github.com/unifiedjs/unified#processing-between-syntaxes), and what you do after to surface results.

**Comment by sebastienbarre** [comment]
Thanks for the quick reply @ChristianMurphy 

I would like to use retext for spell checking and style checking. Retext plugins such as `retext-spell`, `retext-overuse`, `retext-readbility`, `retext-usage`, etc. -- I haven't used any of them yet, but they all look useful.

I'm not entirely sure *where* the errors and warnings would show up--I'm fairly new to this ecosystem--ideally in Visual Studio Code but I'll be fine if they show up in the console/terminal. I'm authoring MDX files from VSCode and launching Next.js in dev mode from a terminal, so I'd assume errors would end up in the latter. 

I gave a shot to the Code Spell Checker extension for VSCode (based on cspell) but its MDX support was less than ideal (it fails to ignore React/JSX components inside MDX). Now if you have good suggestions as to which alternative tool to use to robustly spell check Markdown contents, I'd be very happy to try.

I understand I can't use a retext plugin to further transform text, but the only such plugin I was interested in, `retext-smartypants`, exist as `remark-smartypants`, which I was able to get to work by inserting it in the aforementioned `remarkPlugins` array.

**Comment by ChristianMurphy** [comment, marked as answer]
For grammar checking in CLI or in an editor, I'd recommend using `remark-cli` with `remark-mdx`, like so: https://stackblitz.com/edit/node-rttny7?file=.remarkrc.mjs

<details>

<summary>
<code>.remarkrc.mjs</code> from sandbox
</summary>

```js
import remarkMdx from 'remark-mdx';
import remarkRetext from 'remark-retext';
import { unified } from 'unified';
import retextEnglish from 'retext-english';
import retextSpell from 'retext-spell';
import dictionary from 'dictionary-en';
import retextReadability from 'retext-readability';

const remarkConfig = {
  plugins: [
    remarkMdx,
    [
      remarkRetext,
      unified()
        .use(retextEnglish)
        .use(retextSpell, dictionary)
        .use(retextReadability),
    ],
  ],
};

export default remarkConfig;
```

</details>

which can be used from a command line locally or in a CI server, and can also be integrated with editors like VSCode using the `remark-language-server` (https://github.com/remarkjs/remark-language-server, [docs on integrating with editors](https://github.com/remarkjs/remark-language-server/pull/4))

**Comment by sebastienbarre** [comment]
I see. Thanks for the links and your time. 

Unfortunately I do not control when this happens, per se. When I save my MDX file, Next.js detects the change automatically and triggers its whole machinery to transform that MDX file into a Javascript module that is hot-reloaded and used to update the webapp that ultimately displays the rendered Markdown in a page. I only known that it *does* go through the list of plugins specified in `remarkPlugins` in the process, so if `remark-retext` was there, it would execute it. Now I take it there is no way to instantiate `remark-retext` itself with a list of retext plugins to emulate this chain of `.use(retextEnglish).use(retextSpell, dictionary).use(retextReadability)`.

With that being said, your suggestion is to move this out of the Next.js pipeline, and to try to set it up to run from the CLI to test it out, and ultimately from my IDE. Fair enough :)

Now I was able to replicate the CLI part, so I can technically run `yarn run spell` and it assert that it executes `remark` with the pipeline setup in my `.remarkrc.mjs` -- indeed this showed errors in my MDX files. Again, thank you for helping set that up. This is however not ideal as an experience, right?, I would like the spell-checking to occur straight from my IDE, on the fly. I installed `vs-remark`... but I'm not able to make it work. The "Add examples how to integrate this in editors #4" issue you pointed out leads to [Use `remark-language-server` by remcohaszing · Pull Request #65 · remarkjs/vscode-remark](https://github.com/remarkjs/vscode-remark/pull/65). This looks very promising, but unless I'm mistaken, this was not implemented to completion, or merged in vs-remark main branch? Or am I missing a setting to turn that on? I'm on macOS.

Thank you

**Reply by ChristianMurphy** [reply]
> Unfortunately I do not control when this happens, per se. When I save my MDX file, Next.js detects the change automatically and triggers its whole machinery to transform that MDX file into a Javascript module that is hot-reloaded and used to update the webapp that ultimately displays the rendered Markdown in a page

the same snippet

```js
[
  remarkRetext,
  unified()
    .use(retextEnglish)
    .use(retextSpell, dictionary)
    .use(retextReadability),
]
```

could in theory be dropped into the next configuration as well.
the issue would be that I don't think `next` surfaces the warnings that would be generated.
So the file would be linted, but you wouldn't know what's wrong. 😅 
This is something you could raise with the Next folks, see if they have ideas on how their build system could/would surface warnings.

In the meantime, `remark-cli` offers a way to view the warnings.

> I'm mistaken, this was never implemented correctly, or merged in vs-remark? Or am I missing a setting to turn that on?

`vscode-remark` and `remark-language-server` are both under active development, it appears inline lint warnings is one of the features being developed but not quite released to the vscode extension yet.

**Reply by ProfessorManhattan** [reply]
Hey @ChristianMurphy --- I tested this and it looks like you're right... no errors but no lint messages either

**Reply by ChristianMurphy** [reply]
@ProfessorManhattan I'm assuming your comment relates to https://github.com/orgs/remarkjs/discussions/1137?
You don't appear to be using Next or the language server, I'm not sure what you're testing with, but it's likely not what's being discussed in this thread.

**Comment by sebastienbarre** [comment]
Thanks.

Yes, vs-remark seems to be in rough shape right now for this to work.
As soon as I try to format with `vs-remark`, I get:
```
TypeError: Cannot create property 'rules' on string 'SyntaxError'
```
This was reported as an issue in 2020: [TypeError: Cannot create property 'rules' on string 'SyntaxError' · Issue #18 · remarkjs/vscode-remark](https://github.com/remarkjs/vscode-remark/issues/18)

Unfortunately, it seems that the current [1.3.0] release only supports JSON or YAML, and even the main branch would read my `.remarkrc.mjs` file as JSON, leading to that 'SyntaxError' :( (see [here](https://github.com/remarkjs/vscode-remark/blob/20348d94f8fdff2ddcf74dc7c157b1fa1fe31459/src/utils/workspace.ts)).

Unfortunately I'm not sure how to convert that .mjs file into the equivalent YAML or JSON :(

**Reply by ChristianMurphy** [reply]
The issue you are referencing is specific to YAML https://github.com/remarkjs/vscode-remark/issues/18#issuecomment-650402882 and was solved years ago.

You are working with an `.mjs` file, which is a feature which wasn't added to unified until a year later https://github.com/unifiedjs/unified-engine/pull/50

This should also be resolved by https://github.com/remarkjs/vscode-remark/pull/65
As it upgrades the configuration loader to use `unified-engine` directly.

> I'm not sure how to convert that .mjs file into the equivalent YAML or JSON 

https://github.com/remarkjs/remark/discussions/949

**Comment by sebastienbarre** [comment]
Alright, thanks again for you time. Seems this is an impasse for the time being, as far as the IDE is concerned, but I will subscribe to all these issues and cross my fingers.

And you were right, I was able to drop that pipeline straight into the next configuration array... and as you suspected, it ran fine but never surfaced any warnings.

Now this begs the question... how are people spell-checking their Markdown files nowadays :)
